tests/unit/scripting.tcl

```diff
@@ -1422,6 +1422,18 @@ start_server {tags {"scripting"}} {
             } 0
         }
 
+        # clear the cached server.script_oom just to reproduce issue with EVAL_RO
+        r config set maxmemory 100mb
+        assert_equal [r eval {return 1} 0] 1
+        r config set maxmemory 1
+
+        # Fail to execute regardless of RO script content when we use default flags in OOM condition
+        assert_error {OOM allow-oom flag is not set on the script, can not run it when used memory > 'maxmemory'} {
+            r eval_ro {#!lua flags=
+                return 1
+            } 0
+        }
```

Comment:
All write commands are implicitly use-memory, and since they are disallowed in EVAL*_RO I would expect that an EVAL*_RO to never fail on OOM conditions.

Reply:
that's right. but the shebang flags can define that the script should not be run in OOM condition.
do you think we should ignore that flag in EVAL_RO, FCALL_RO and let the script run even when we're in OOM state?
@MeirShpilraien @yoav-steinberg WDYT?

Reply:
thinking about it more, a script that declares a shebang, has to declare the `no-writes` in order to be used in EVAL_RO.
that means that the entity that defines when the script will fail, is the script author, not the caller (i.e. the script code not which command was used).
this is certainly true for functions where the caller doesn't necessarily knows what the script really does.

so i think we should focus on two things:
1. it's quite clear what should be done with functions, and i suppose we want eval to be consistent with that.
2. i suppose the only debate here should be about eval scripts that don't declare a shebang at all.

i think these work just fine in both EVAL and EVAL_RO. i'll add a test to demonstrate that.

Reply:
OOM doesn't really make sense in the context of EVAL_RO scripts though. It seems like an unexpected failure mode to have a script, which doesn't use memory, to fail on OOM conditions if you add a shebang. Can't we just document that these two modes are different?

Reply:
document what?
that in both EVAL_RO and FCALL_RO, the default flags are ignored and the script is executed as if it declared `allow-oom`?

would you consider doing the same for `no-writes`?
i.e. user will be able to call a function that didn't declare `no-writes` by using FCALL_RO?
the script would start running, and then abort with error if it uses any write command?

Reply:
I'm suggesting the default flags can be different between EVAL and EVAL_RO (and their FCALL counterparts) when a shebang is provided. This different default behavior is consistent with my expectation of what EVAL_RO should do.

I think the alternative you're proposing is that to use a script with EVAL_RO or FCALL_RO, it must have the `no-write` and `allow-oom` flags declared, since that is the only configuration that logically makes sense. I think this is a little bit convoluted, but I also think this proposal is valid.

Reply:
so we have two alternatives, and you're saying both are ok with you?

1. scripts (with or without shebang, and even functions), have different default flags when used in the _RO command variant.
2. scripts that use shebang must declare no-write (and possibly allow-oom) in order to be used in the _RO commmand variants. (scripts without shebang don't need that)

you said yourself that at least for functions, the first option is sub-optimal, and i agree, it's very odd since the person implementing the function and the one who loaded it are different than the one who calls it, and might have no clue what it actually does.

you said that the second proposal (that's what's implemented) is valid, right? so can we agree on the second option?
@MeirShpilraien @yossigo please comment.

Reply:
I think the second option is better and I think this is what was decided back then when we develop it, we said that when eval script is used with shebang the behaviour will match the function flags, and on function you must specify `allow-oom` in order to run it when OOM reached (even if used with `fcall_ro`).

Reply:
I discussed this with Yossi, we argue that as soon as there's shebang, the intention of the script is explicit, and it is consistent (no matter which command).

so i think we all agree on option [2] listed above, which is what the code implements already.
so @madolson please ack and we'll merge this (unrelated) bugfix and move on.

Reply:
I'll still argue that EVAL_RO and FCALL_RO should also require "allow-oom" to run, in the same way we require them to have the "allow-write" flag. Which is not the current implementation, unless I really misunderstand the side effects of setting the flag.

Reply:
you mean that any attempt to run a function with FCALL_RO should fail unless it is marked with `allow-oom` (even when we're not in OOM state)?

i think that would be suitable for a command named FCALL_OOM, buy luckily we don't have / need one.
i kinda regret adding that command in the first place (EVAL_RO). if it's sole purpose is for client routing to replicas, and the grantee that no matter the script it cannot do any writes, i don't see why it should require the `allow-oom` flag.

can you explain again what's the benefit of doing that?

Reply:
Maybe a simpler idea is that `no-write` implies `allow-oom`? I'm not really aware of a case where a non-write command would consume additional memory, so there isn't really a reason to block in OOM scenarios. Since OOM scenarios are infrequent, I would rather guide users to setting the right flags so their application works as expected in OOM scenarios.

Reply:
we discussed that in a core-team meeting and concluded we wanna proceed with this idea (`no-write` to imply `allow-oom`), i'll make a separate PR for that and we can further discuss it there.

Reply:
on a second thought, although that change is completely unrelated to the other fix in this PR, effects and the tests might be related, so let's keep the discussion here.

So some background:
in #10066 we added the `no-write` and `allow-oom` flags, i seem to remember that we did consider setting allow-oom implicitly when `no-write` is set, it was discussed here https://github.com/redis/redis/pull/10066#discussion_r781809404, and was rejected because i wanted similar flags and behavior as normal redis commands and module commands (to avoid confusion).
but later we negated the flags so we have different defaults, so maybe we should reconsider that (despite being a breaking change comparing to redis 7.0.0)

@MeirShpilraien do you remember any other detail?

Reply:
For the record, i discussed it with Meir.
It seems we clearly stated in the original PR top comment that:

> if the function will not turn on this flag it will not be possible to run it if OOM reached (even if the function declares no-writes and even if fcall_ro is used). If this flag is set, any command will be allow on OOM (even those that is marked with CMD_DENYOOM)

But looking back at the original discussions, we can't spot any reason not to revert that decision and have no-writes implicitly count as allow-oom too.